util/util.go

```diff
@@ -0,0 +1,232 @@
+
+	// todo: change environment variables into configuration
+	// add handling functionality
+	certificateProps := certificate{
+		cPath:    			os.Getenv("CERT"),
+		cKeyPath: 			os.Getenv("CERT_KEY"),
+		caPath:             os.Getenv("CA"),
+	}
+
+	certificate, err := tls.LoadX509KeyPair(certificateProps.cPath, certificateProps.cKeyPath)
+	if err != nil {
+		return nil,fmt.Errorf("could not load server key pair: %s", err)
+	}
+
+	// Create a certificate pool from the certificate authority
+	certPool := x509.NewCertPool()
+	ca, err := ioutil.ReadFile(certificateProps.caPath)
+	if err != nil {
+		return nil, fmt.Errorf("could not read ca certificate: %s", err)
+	}
+	// CA file for let's encrypt is located under domain conf as `chain.pem`
+	// pass chain.pem location
+	// Append the client certificates from the CA
+	if ok := certPool.AppendCertsFromPEM(ca); !ok {
+		return nil, errors.New("failed to append client certs")
+	}
+
+	// Create the TLS credentials
+	creds := credentials.NewTLS(&tls.Config{
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+		Certificates: []tls.Certificate{certificate},
+		ClientCAs:    certPool,
+	})
+	return creds, nil
+}
+
+func (s server) GrpcOpts() ([]grpc.ServerOption, error) {
+
+	if s.tls {
+		creds,err := GetCreds()
+
+		if err != nil {
+			return []grpc.ServerOption{}, errors.New("Error on retrieving certificates: "+err.Error())
+		}
+		return []grpc.ServerOption{grpc.Creds(creds)}, nil
+	}
+	return []grpc.ServerOption{}, nil
+}
+
+func (s server) GetGRPCServer(opts ...grpc.ServerOption) *grpc.Server {
+
+	streamInterceptor := func(srv interface{}, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		if err := s.auth.AuthenticateContext(stream.Context()); err != nil {
+			return err
+		}
+		return handler(srv, stream)
+	}
+
+	unaryInterceptor := func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		if err := s.auth.AuthenticateContext(ctx); err != nil {
+			return nil, err
+		}
+		return handler(ctx, req)
+	}
+
+	opts = append([]grpc.ServerOption{
+		grpc.StreamInterceptor(streamInterceptor),
+		grpc.UnaryInterceptor(unaryInterceptor),
+	}, opts...)
+	return grpc.NewServer(opts...)
+}
+
+func readContent(path string) error {
+	cont, err := os.Open(path)
+	if err != nil {
+		log.Fatal(err)
+		return err
+	}
+	_, err = io.Copy(os.Stdout, cont)
+	if err != nil {
+		log.Fatal(err)
+		return err
+	}
+	return nil
+}
+
+
+
+func InitilizegRPCServer() *server {
+
+
+	store, err := database.NewStore()
+
+	if err != nil {
+		log.Fatalf("failed to connect to database: %v", err)
+	}
+	// todo: change handling of tls, the example below not good enough
+
+	tls := true
+	mode := os.Getenv("SSL_OFF")
+	if strings.ToLower(mode) == "true" {
+		tls = false
+	}
+
+	s := &server{
+		store:   store,
+		auth:    NewAuthenticator(os.Getenv("SIGNIN_KEY")),
+		tls:     tls,
+	}
```

Comment:
Although the name of the function looks very serious, the content of the function looks ugly, handling TLS can be much better and this function either removed or optimized with better one.

Reply:
it can be handled in the configuration file